src/cpp/rtps/reader/StatefulReader.cpp

```diff
@@ -862,22 +880,45 @@ bool StatefulReader::change_removed_by_history(
 {
     std::lock_guard<RecursiveTimedMutex> guard(mp_mutex);
 
-    if (is_alive_ && a_change->is_fully_assembled())
+    if (is_alive_)
     {
-        if (wp != nullptr || matched_writer_lookup(a_change->writerGUID, &wp))
+        if (a_change->is_fully_assembled())
         {
-            wp->change_removed_from_history(a_change->sequenceNumber);
-        }
+            if (wp != nullptr || matched_writer_lookup(a_change->writerGUID, &wp))
+            {
+                wp->change_removed_from_history(a_change->sequenceNumber);
+            }
+
+            if (!a_change->isRead &&
+                    get_last_notified(a_change->writerGUID) >= a_change->sequenceNumber)
+            {
+                if (0 < total_unread_)
+                {
+                    --total_unread_;
+                }
 
-        if (!a_change->isRead &&
-                get_last_notified(a_change->writerGUID) >= a_change->sequenceNumber)
+            }
+        }
+        else
         {
-            if (0 < total_unread_)
+            /* A not fully assembled fragmented sample may be removed due to being received a newer sample and KEEP_LAST
+             * policy. In this case WriterProxy should consider it as irrelevant for avoiding enter in a loop asking for
+             * it.
```

Comment:
```suggestion
            /* A not fully assembled fragmented sample may be removed when receiving a newer sample and KEEP_LAST
             * policy. The WriterProxy should consider it as irrelevant to avoid an infinite loop asking for it.
```